harness/tests/determined/pytorch/experimental/test_torch_batch_process.py

```diff
@@ -0,0 +1,570 @@
+import math
+from typing import Any, Dict, List, Optional, Type
+from unittest.mock import MagicMock, Mock, call, patch
+
+import pytest
+import torch
+from torch.utils.data import Dataset
+
+from determined import pytorch
+from determined.pytorch.experimental._torch_batch_process import (
+    TorchBatchProcessor,
+    TorchBatchProcessorContext,
+    get_default_device,
+    torch_batch_process,
+)
+
+DEFAULT_SLOT_IDS = [0]
+DEFAULT_ADDRS = ["0.0.0.12"]
+DEFAULT_DEVICE = torch.device("cuda", 1)
+DEFAULT_STORAGE_PATH = "default_storage_path"
+
+
+@pytest.fixture
+def batch_processor_with_avg_metric_reducer() -> Type[TorchBatchProcessor]:
+    class MySumMetricReducer(pytorch.MetricReducer):
+        def __init__(self) -> None:
+            self.reset()
+
+        def reset(self) -> None:
+            self.sum = 0
+
+        def update(self, value: List[int]) -> None:
+            self.sum += sum(value)
+
+        def per_slot_reduce(self) -> int:
+            return self.sum
+
+        def cross_slot_reduce(self, per_slot_metrics: List[int]) -> int:
+            return sum(per_slot_metrics)
+
+    class MyProcessor(TorchBatchProcessor):
+        def __init__(self, context: TorchBatchProcessorContext) -> None:
+            self.context = context
+            self.reducer_sum = context.wrap_reducer(reducer=MySumMetricReducer(), name="sum_metric")
+
+        def process_batch(self, batch: Any, batch_idx: int) -> None:
+            self.reducer_sum.update(batch)
+
+    return MyProcessor
+
+
+def _get_index_dataset(data_length: int = 50) -> Dataset:
+    class IndexData(Dataset):
+        def __init__(self, data_length: int) -> None:
```

Comment:
similar, I think you could just define IndexDataset directly, the getter function for a class seems like an unnecessary indirection.